lms/djangoapps/mobile_api/users/serializers.py

```diff
@@ -141,6 +186,123 @@ class Meta:
         lookup_field = 'username'
 
 
+class CourseEnrollmentSerializerModifiedForPrimary(CourseEnrollmentSerializer):
+    """
+    Serializes CourseEnrollment models for API v4.
+
+    Adds `course_status` field into serializer data.
+    """
+
+    course_status = serializers.SerializerMethodField()
+    course_progress = serializers.SerializerMethodField()
+    course_assignments = serializers.SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.course = modulestore().get_course(self.instance.course.id)
+
+    def get_course_status(self, model: CourseEnrollment) -> Optional[Dict[str, List[str]]]:
+        """
+        Gets course status for the given user's enrollments.
+        """
+        try:
+            block_id = str(get_key_to_last_completed_block(model.user, model.course.id))
+        except UnavailableCompletionData:
+            block_id = ''
+
+        if not block_id:
+            return None
+
+        path, unit_name = self._get_last_visited_block_path_and_unit_name(block_id)
+        if not path and unit_name:
+            return None
+
+        path_ids = [str(block.location) for block in path]
+
+        return {
+            'last_visited_module_id': path_ids[0],
+            'last_visited_module_path': path_ids,
+            'last_visited_block_id': block_id,
+            'last_visited_unit_display_name': unit_name,
+        }
+
+    @staticmethod
+    def _get_last_visited_block_path_and_unit_name(
+        block_id: str
+    ) -> Union[Tuple[None, None], Tuple[List['XBlock'], str]]:  # noqa: F821
+        """
+        Returns the path to the latest block and unit name visited by the current user.
+        """
+        try:
+            last_visited_block = modulestore().get_item(UsageKey.from_string(block_id))
+            vertical = last_visited_block.get_parent()
+            sequential = vertical.get_parent()
+            chapter = sequential.get_parent()
+            course = chapter.get_parent()
+        except (ItemNotFoundError, InvalidKeyError, AttributeError):
+            return None, None
+
+        path = [sequential, chapter, course]
+
+        return path, vertical.display_name
+
+    def get_course_progress(self, model: CourseEnrollment) -> Dict[str, int]:
+        """
+        Returns the progress of the user in the course.
+        """
+        return self.calculate_progress(model)
+
+    def get_course_assignments(self, model: CourseEnrollment) -> Dict[str, Optional[List[Dict[str, str]]]]:
```

Comment:
I would like this logic to exist in a central platform Python API instead of the Mobile API. If the Learning MFE or Learner Dash MFE ever needs this same data, they should be able to call this function instead of reimplementing it. Could you move it? Maybe lms/djangoapps/courseware/courses.py is the right place?

In general, when implementing Mobile API features, try to keep the student/grading/progress/access logic in core applications, and then just use the mobile_api as a thin wrapper for presenting that information to the mobile apps. That way, both the Web and Mobile versions of the platform can have efficient behavior and we will have less business logic to test.

Reply:
@kdmccormick I agree, it sounds logical. I moved `get_course_assignments` here: `lms/djangoapps/courseware/courses.py`